File: app/src/main/java/com/friday/ar/service/AccountSyncService.java
package com.friday.ar.service;

import android.app.DownloadManager;
import android.app.job.JobParameters;
import android.app.job.JobService;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.util.Log;

import com.friday.ar.util.FileUtil;
import com.friday.ar.util.UserUtil;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;

import java.io.File;
import java.io.IOException;

public class AccountSyncService extends JobService {
    private static final String LOGTAG = "SynchronizationService";

    public AccountSyncService() {
    }

    @Override
    public boolean onStartJob(JobParameters jobParameters) {
        Log.d(LOGTAG, "started synchronization service");
        FirebaseUser firebaseUser = FirebaseAuth.getInstance().getCurrentUser();
        UserUtil util = new UserUtil(getApplicationContext());
        if (firebaseUser == null) {
            jobFinished(jobParameters, false);
            return false;
        }
        if (firebaseUser.getPhotoUrl() != null) {
            File avatarFile = util.getAvatarFile();
            avatarFile.delete();
            DownloadManager.Request request = new DownloadManager.Request(firebaseUser.getPhotoUrl());
            request.setNotificationVisibility(DownloadManager.Request.VISIBILITY_HIDDEN)
                    .setDestinationInExternalFilesDir(getApplicationContext(), "profile", File.separator + "avatar.jpg")
                    .setVisibleInDownloadsUi(false);
            registerReceiver(new BroadcastReceiver() {
                @Override
                public void onReceive(Context context, Intent intent) {
                    context.unregisterReceiver(this);
                    Log.d(LOGTAG, "synchronized account avatar");
                    try {
                        FileUtil.moveFile(new File(getExternalFilesDir("profile") + File.separator + "avatar.jpg"), avatarFile);
                    } catch (IOException e) {
                        Log.e(LOGTAG, e.getLocalizedMessage(), e);
                    } finally {
                        jobFinished(jobParameters, false);
                    }
                }
            }, new IntentFilter(DownloadManager.ACTION_DOWNLOAD_COMPLETE));
            DownloadManager dm = (DownloadManager) getSystemService(DOWNLOAD_SERVICE);
            dm.enqueue(request);
        }
        return true;
    }

    @Override
    public boolean onStopJob(JobParameters jobParameters) {
        Log.d(LOGTAG, "finished synchronization service");
        return false;
    }
}

File: app/src/main/java/com/friday/ar/service/AccountSyncService.kt
package com.friday.ar.service

import android.app.DownloadManager
import android.app.Service
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.os.Binder
import android.os.IBinder
import android.util.Log
import com.friday.ar.Constant
import com.friday.ar.util.UserUtil
import com.google.firebase.auth.FirebaseAuth
import java.io.IOException
import java.nio.file.Files
import java.nio.file.StandardCopyOption

class AccountSyncService : Service() {
    companion object {
        private const val LOGTAG = "SynchronizationService"
    }

    override fun onBind(p0: Intent?): IBinder? {
        Log.d(LOGTAG, "started synchronization service")
        val firebaseUser = FirebaseAuth.getInstance().currentUser
        if (firebaseUser != null && firebaseUser.photoUrl != null) {
            UserUtil(applicationContext).avatarFile.delete()
            val request = DownloadManager.Request(firebaseUser.photoUrl)
            request.setNotificationVisibility(DownloadManager.Request.VISIBILITY_HIDDEN)
                    .setDestinationInExternalFilesDir(applicationContext, "profile", "/avatar.jpg")
                    .setNotificationVisibility(DownloadManager.Request.VISIBILITY_HIDDEN)
            registerReceiver(downloadFinishReceiver, IntentFilter(DownloadManager.ACTION_DOWNLOAD_COMPLETE))
            val dm = getSystemService(Context.DOWNLOAD_SERVICE) as DownloadManager
            dm.enqueue(request)
        }
        return SyncServiceBinder()
    }

    private val downloadFinishReceiver = object : BroadcastReceiver() {
        override fun onReceive(context: Context, intent: Intent) {
            context.unregisterReceiver(this)
            val util = UserUtil(applicationContext)
            val avatarFile = util.avatarFile
            Log.d(LOGTAG, "synchronized account avatar")
            try {
                Files.move(getExternalFilesDir("profile/avatar.jpg")!!.toPath(), avatarFile.toPath(), StandardCopyOption.REPLACE_EXISTING)
                val avatarSyncedIntent = Intent()
                avatarSyncedIntent.action = Constant.BroadcasteceiverActions.BROADCAST_ACCOUNT_SYNCED
                sendBroadcast(avatarSyncedIntent)
            } catch (e: IOException) {
                Log.e(LOGTAG, e.localizedMessage, e)
            }
        }
    }

    inner class SyncServiceBinder : Binder() {
        fun getService(): AccountSyncService = this@AccountSyncService
    }
}
